// feishu-docx/dist/renderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trimLastNewline = exports.Renderer = void 0;
class Renderer {
    constructor(doc, options = {}) {
        var _a, _b;
        this.blockMap = {};
        this.fileTokens = {};
        this.outputUnsupported = false;
        const { debug = false, outputUnsupported } = options;
        this.documentId = ((_a = doc === null || doc === void 0 ? void 0 : doc.document) === null || _a === void 0 ? void 0 : _a.document_id) || '';
        this.fileTokens = {};
        (_b = doc === null || doc === void 0 ? void 0 : doc.blocks) === null || _b === void 0 ? void 0 : _b.forEach((block) => {
            this.blockMap[block === null || block === void 0 ? void 0 : block.block_id] = block;
        });
        this.debug = debug;
        this.outputUnsupported = outputUnsupported;
    }
    /**
     * Parse Feishu doc to new format
     * @returns Text of new format content.
     */
    parse() {
        const entryBlock = this.blockMap[this.documentId];
        return this.parseBlock(entryBlock, 0);
    }
    parseBlock(block, indent) {
        throw new Error('Not implemented');
    }
    /**
     * Add a file token to context
     * @param type
     * @param token
     */
    addFileToken(type, token) {
        this.fileTokens[token] = {
            token,
            type,
        };
    }
}
exports.Renderer = Renderer;
/**
 * 去掉末尾 1 个换行
 * @param str
 * @returns
 */
const trimLastNewline = (str) => {
    return str.replace(/\n$/, '');
};
exports.trimLastNewline = trimLastNewline;
